UefiPayloadPkg/SPI/fch_spi_ctrl.c

```diff
@@ -0,0 +1,277 @@
+#include <Include/Library/DebugLib.h>
+#include <Include/PiDxe.h>
+
+#include "kconfig.h"
+#include "SPI.h"
+#include "utils.h"
+#include "stopwatch.h"
+#include "fch_spi_util.h"
+#include "SPIgeneric.h"
+#include "lpc.h"
+#include "pci_devs.h"
+#include "spi_flash.h"
+#include "pci_ops.h"
+
+/* SPDX-License-Identifier: GPL-2.0-only */
+
+#define GRANULARITY_TEST_4k		0x0000f000		/* bits 15-12 */
+#define WORD_TO_DWORD_UPPER(x)		((x << 16) & 0xffff0000)
+
+/* SPI MMIO registers */
+#define SPI_RESTRICTED_CMD1		0x04
+#define SPI_RESTRICTED_CMD2		0x08
+#define SPI_CNTRL1			0x0c
+#define SPI_CMD_CODE			0x45
+#define SPI_CMD_TRIGGER			0x47
+#define   SPI_CMD_TRIGGER_EXECUTE	BIT7
+#define SPI_TX_BYTE_COUNT		0x48
+#define SPI_RX_BYTE_COUNT		0x4b
+#define SPI_STATUS			0x4c
+#define   SPI_DONE_BYTE_COUNT_SHIFT	0
+#define   SPI_DONE_BYTE_COUNT_MASK	0xff
+#define   SPI_FIFO_WR_PTR_SHIFT		8
+#define   SPI_FIFO_WR_PTR_MASK		0x7f
+#define   SPI_FIFO_RD_PTR_SHIFT		16
+#define   SPI_FIFO_RD_PTR_MASK		0x7f
+
+static VOID dump_state(CONST char *str, UINT8 phase)
+{
+	UINT8 dump_size;
+	UINT32 addr;
+
+	if (!CONFIG(SOC_AMD_COMMON_BLOCK_SPI_DEBUG))
+		return;
+	DEBUG((EFI_D_INFO, "%a: SPI: %s\n", __FUNCTION__, str));
+	DEBUG((EFI_D_INFO, "%a: Cntrl0: %x\n", __FUNCTION__, spi_read32(SPI_CNTRL0)));
+	DEBUG((EFI_D_INFO, "%a: Status: %x\n", __FUNCTION__, spi_read32(SPI_STATUS)));
+
+	addr = spi_get_bar() + SPI_FIFO;
+	if (phase == 0) {
+		dump_size = spi_read8(SPI_TX_BYTE_COUNT);
+		DEBUG((EFI_D_INFO, "%a: TxByteCount: %x\n", __FUNCTION__, dump_size));
+		DEBUG((EFI_D_INFO, "%a: CmdCode: %x\n", __FUNCTION__, spi_read8(SPI_CMD_CODE)));
+	} else {
+		dump_size = spi_read8(SPI_RX_BYTE_COUNT);
+		DEBUG((EFI_D_INFO, "%a: RxByteCount: %x\n", __FUNCTION__, dump_size));
+		addr += spi_read8(SPI_TX_BYTE_COUNT);
+	}
+
+	if (dump_size > 0) {
+    //hexdump((VOID *)addr, dump_size);
+    DEBUG((EFI_D_INFO, "%a: HEXDUMPING NOT IMPLEMENTED: %x\n", __FUNCTION__, dump_size));
+  }
+}
+
+static int wait_for_ready(VOID)
+{
+	CONST UINT32 timeout_ms = 500;
+	struct stopwatch sw;
+
+	stopwatch_init_msecs_expire(&sw, timeout_ms);
```

Comment:
this is coreboot specific. You should replace it with EDK2 implementation of timer. UEFIPayloadPkg uses ACPITimer:
`UefiPayloadPkg/Library/AcpiTimerLib/AcpiTimerLib.c`. It would be easier to remove the whole stopwatch, and jsut use the delay from UefiPayloadPkg/Library/AcpiTimerLib/AcpiTimerLib.c in a loop until your reach the specified timeout.

Reply:
@IgorBagnucki

Reply:
using AcpiTimerLib.c from now